### PR TITLE
Some fixes for `stake` commands

### LINKED ISSUE #1657: `nucypher stake collect-reward` doesn't work with `--staking-address` flag
I'm trying to withdraw my rewards with the command `nucypher stake collect-reward --staking-reward --policy-reward --staking-address 0xf56727fc21bf5f1f7335c71df22e68a269283df7`

```
The Holder of Stakes.                                      

Traceback (most recent call last):
  File "/home/ilhan/nucypher-venv/bin/nucypher", line 8, in <module>
    sys.exit(nucypher_cli())
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/cli/options.py", line 134, in wrapper
    return func(**kwargs)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/cli/options.py", line 134, in wrapper
    return func(**kwargs)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/cli/options.py", line 134, in wrapper
    return func(**kwargs)
  [Previous line repeated 1 more time]
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/cli/commands/stake.py", line 745, in collect_reward
    STAKEHOLDER = transacting_staker_options.create_character(emitter, config_file)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/cli/commands/stake.py", line 211, in create_character
    initial_address=initial_address)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/cli/commands/stake.py", line 150, in create_character
    individual_allocation=individual_allocation)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/config/node.py", line 287, in produce
    character = self.CHARACTER_CLASS(**merged_parameters)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/blockchain/eth/actors.py", line 1341, in __init__
    self.assimilate(checksum_address=initial_address, password=password)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/blockchain/eth/decorators.py", line 70, in wrapped
    return func(*args, **kwargs)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/blockchain/eth/actors.py", line 1365, in assimilate
    self.wallet.activate_account(checksum_address=checksum_address, password=password)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/blockchain/eth/decorators.py", line 70, in wrapped
    return func(*args, **kwargs)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/blockchain/eth/actors.py", line 1305, in activate_account
    transacting_power.activate(password=password)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/crypto/powers.py", line 160, in activate
    self.unlock_account(password=password)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/crypto/powers.py", line 180, in unlock_account
    unlocked = self.blockchain.client.unlock_account(address=self.account, password=password, duration=duration)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/nucypher/blockchain/eth/clients.py", line 332, in unlock_account
    return self.w3.geth.personal.unlockAccount(address, password, duration)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/web3/module.py", line 43, in caller
    result = w3.manager.request_blocking(method_str, params, error_formatters)
  File "/home/ilhan/nucypher-venv/lib/python3.6/site-packages/web3/manager.py", line 152, in request_blocking
    raise ValueError(response["error"])
ValueError: {'code': -32000, 'message': 'could not decrypt key with given password'}
Sentry is attempting to send 0 pending error messages
Waiting up to 2.0 seconds
Press Ctrl-C to quit
```

I have also tried to collect rewards with the following flags: 

`--withdraw-address 0xf56727fc21bf5f1f7335c71df22e68a269283df7`
`--no-hw-wallet`

With both flags the same error comes up.

### LINKED ISSUE #1696: `actions.select_stake` incorrectly shows divisible sub-stakes
`painting.paint_stakes` prints all sub-substakes instead of filtered divisible sub-stakes from `stakes` variable

### LINKED ISSUE #2113: Only able to divide stake when staking address is explicitly provided
**Describe the Bug**
I encountered an interesting bug when using `nucypher stake divide`. When running the command without specifying my staking address I get the following:
```
$ nucypher stake divide
	Account
--  ------------------------------------------
 0  0x63e478bc474eBb6c31568ff131cCd95C24bfD552
 1  0x270b3f8af5ba2B79ea3Bd6a6Efc7ecAB056d3E3f
 2  0x45D33d1Ff0A7E696556f36DE697E5C92C2CCcFaE
Select index of staking account [0]: 1
Selected 1: 0x270b3f8af5ba2B79ea3Bd6a6Efc7ecAB056d3E3f
NOTE: Showing divisible stakes only
No divisible stakes found.
```

However if I specify the same staking address explicitly as a parameter then I'm allowed to divide the stake:
```
(nucypher) derek@derek-home-server:~/nucypher/nucypher$ nucypher stake divide --staking-address 0x270b3f8af5ba2B79ea3Bd6a6Efc7ecAB056d3E3f
NOTE: Showing divisible stakes only

Network Ibex ═════════════════════════════════════════
Staker 0x270b3f8af5ba2B79ea3Bd6a6Efc7ecAB056d3E3f ════
Worker 0x45D33d1Ff0A7E696556f36DE697E5C92C2CCcFaE ════
--------------  ----------------
Status      	Committed #18436
Restaking   	Yes (Unlocked)
Winding Down	No
Unclaimed Fees  0 ETH
Min fee rate	0 ETH
--------------  ----------------
╒═══════╤═════════════════════════════╤═════════════╤═════════════╤═══════════════╕
│   Idx │ Value                   	│   Remaining │ Enactment   │ Termination   │
╞═══════╪═════════════════════════════╪═════════════╪═════════════╪═══════════════╡
│ 	0 │ 31437.841006996376688377 NU │      	31 │ Jun 19 2020 │ Jul 22 2020   │
╘═══════╧═════════════════════════════╧═════════════╧═════════════╧═══════════════╛
Select Stake: 0
Enter target value (15000 NU - 16437.841006996376688377 NU): 15000
Enter number of periods to extend: 20

══════════════════════════════ ORIGINAL STAKE ════════════════════════════

Staking address: 0x270b3f8af5ba2B79ea3Bd6a6Efc7ecAB056d3E3f
~ Original Stake: | - | 0x270b | 0x45D3 | 0 | 31437.841006996376688377 NU | 33 periods . | Jun 19 20:00 EDT - Jul 22 20:00 EDT


══════════════════════════════ STAGED STAKE ══════════════════════════════

Staking address: 0x270b3f8af5ba2B79ea3Bd6a6Efc7ecAB056d3E3f
~ Chain  	-> ID # 4 | Rinkeby
~ Value  	-> 15000 NU (15000000000000000000000 NuNits)
~ Duration   -> 53 Days (53 Periods)
~ Enactment  -> Jun 19 20:00 EDT (period #18433)
~ Expiration -> Aug 11 20:00 EDT (period #18486)
    
═════════════════════════════════════════════════════════════════════════
Publish stake division to the blockchain? [y/N]: y
Enter password to unlock account 0x270b3f8af5ba2B79ea3Bd6a6Efc7ecAB056d3E3f:
Confirm transaction DIVIDESTAKE on hardware wallet... (76058 gwei @ 1000000000)
Broadcasting DIVIDESTAKE Transaction (76058 gwei @ 1000000000)...
Successfully divided stake
OK | 0x74ddd647de6eaca7ef0c485706ef526001d959a3c2eaa98699e087a7d259d08b (75349 gas)
Block #6711982 | 0xd1c6d6df257ecd05632550565edb709ae577066a60ca433bc4d23de5fb332009
 See https://rinkeby.etherscan.io/tx/0x74ddd647de6eaca7ef0c485706ef526001d959a3c2eaa98699e087a7d259d08b
...
```

My stakeholder configuration:
```
(nucypher) derek@derek-home-server:~/nucypher/nucypher$ nucypher stake config
Stakeholder Configuration /home/derek/.local/share/nucypher/stakeholder.json
 =======================================================
{
  "provider_uri": “<infura>",
  "poa": null,
  "light": null,
  "domains": [
    "ibex"
  ],
  "checksum_addresses": [],
  "signer_uri": "clef:///home/derek/.clef/clef.ipc",
  "version": 1
}
```

Also, there is a bug in the string parameter we pass to our `PROMPT_STAKE_DIVIDE_VALUE` template string:
```
  File "/home/derek/nucypher/nucypher/nucypher/cli/commands/stake.py", line 664, in divide
	prompt = PROMPT_STAKE_DIVIDE_VALUE.format(minimm=min_allowed_locked, maximum=str(max_divide_value))
KeyError: 'minimum'
```
i.e. `minimm` should be `minimum`. I manually changed it locally to proceed.

**To Reproduce**
See above.

**Traceback or Screenshots (Optional) **
See above

**System (please complete the following information):**
 - OS Version: Linux Ubuntu 18.04.4
 - Nucypher Version: `master` @ https://github.com/nucypher/nucypher/commit/202e502ad4239227b613c1346cd161723d892765
 - Ethereum Node Version:  Geth 1.9.15-stable-0f77f34b

### LINKED ISSUE #2116: incorrect work "nucypher stake prolong"
Run `nucypher stake prolong` command
An error occurs at the `Select Stake` step.

Stake prolong

 - OS Version: Ubuntu   18.04.4 LTS
 - Nucypher Version: version 2.1.0-beta.12
 - Ethereum Node Version:  Geth 1.9.15-stable

### PR DESCRIPTION
**Pull request by vzotova**
Fixes #2113
Fixes #2116
Fixes #1657 
Fixes #1696 

* Adds status for sub-stake
* Simplifies sub-stake selection
* Splitted `StakeHolder.assimilate()` into two methods 
* More tests

### CODE REVIEW

File: nucypher/cli/commands/stake.py [hunk 1]
```diff
@@ -684,15 +694,15 @@ def divide(general_config, transacting_staker_options, config_file, force, value
 
     # Authenticate
     password = transacting_staker_options.get_password(blockchain, client_account)
+    STAKEHOLDER.assimilate(password=password)
 
     # Consistency check to prevent the above agreement from going stale.
     last_second_current_period = STAKEHOLDER.staking_agent.get_current_period()
     if action_period != last_second_current_period:
-        emitter.echo(PERIOD_ADVANCED_WARNING, red='red')
+        emitter.echo(PERIOD_ADVANCED_WARNING, color='red')
```
Comment (derekpierre) [inline nucypher/cli/commands/stake.py:702]: 😱

File: nucypher/blockchain/eth/actors.py [hunk 2]
```diff
@@ -851,7 +854,6 @@ def to_dict(self) -> dict:
     @property
     def is_staking(self) -> bool:
         """Checks if this Staker currently has active stakes / locked tokens."""
-        self.stakes.refresh()
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/actors.py:854]: I'm trying to make abstraction more clear. `Staker` methods assume that `stakes` are already refreshed - because refreshing is heavy and should be controlled outside

File: nucypher/blockchain/eth/actors.py [hunk 3]
```diff
@@ -874,31 +875,20 @@ def current_stake(self) -> NU:
...
-        result = stake.value >= min_divisible_stake_value
-        return result
+        stakes = (stake for stake in self.stakes if stake.status(staker_info, current_period).is_child(parent_status))
+        return stakes
 
-    @property
-    def divisible_stakes(self) -> List[Stake]:
-        """chop chop"""
-        stakes = list(filter(self.__filter_divisible_stakes, self.stakes))
+    def sorted_stakes(self, parent_status: Stake.Status = None) -> List[Stake]:
```
Comment (KPrasch) [inline nucypher/blockchain/eth/actors.py:888]: :+1:

File: nucypher/blockchain/eth/token.py [hunk 4]
```diff
@@ -159,6 +161,24 @@ class Stake:
...
+        """
+        Sub-stake status.
+        """
+        INACTIVE = 1   # Unlocked inactive
+        UNLOCKED = 2   # Unlocked active
+        LOCKED = 3     # Locked but not editable
+        EDITABLE = 4   # Editable
+        DIVISIBLE = 5  # Editable and divisible
+
+        def is_child(self, other: 'Status') -> bool:
```
Comment (KPrasch) [inline nucypher/blockchain/eth/token.py:174]: :+1: - Excellent

File: nucypher/blockchain/eth/token.py [hunk 5]
```diff
@@ -249,43 +269,56 @@ def address_index_ordering_key(self):
     # Metadata
     #
 
-    @property
-    def is_expired(self) -> bool:
-        current_period = self.staking_agent.get_current_period()  # TODO #1514 this is online only.
-        return bool(current_period > self.final_locked_period)
+    def status(self, staker_info: StakerInfo = None, current_period: Period = None) -> Status:
+        if self._status:
```
Comment (KPrasch) [inline nucypher/blockchain/eth/token.py:273]: Perhaps a docstring here describing the status tree.

File: nucypher/blockchain/eth/token.py [hunk 6]
```diff
@@ -383,14 +416,13 @@ def sync(self) -> None:
...
+        if not self.first_locked_period == stake_info.first_period:
             raise self.StakingError("Inconsistent staking cache.  Make sure your node is synced and try again.")
 
         # Mutate the instance with the on-chain values
-        self.final_locked_period = last_period
-        self.value = NU.from_nunits(locked_value)
-        self.worker_address = self.staking_agent.get_worker_from_staker(staker_address=self.staker_address)
+        self.final_locked_period = stake_info.last_period
+        self.value = NU.from_nunits(stake_info.locked_value)
+        self._status = None
```
Comment (KPrasch) [inline nucypher/blockchain/eth/token.py:425]: Does this indicate an *unknown* status?

File: nucypher/blockchain/eth/token.py [hunk 7]
```diff
@@ -491,8 +525,10 @@ def initialize_stake(cls, staker, amount: NU, lock_periods: int) -> 'Stake':
 
     def prolong(self, additional_periods: int):
         self.sync()
-        if self.is_expired:
-            raise self.StakingError(f'Cannot divide an expired stake. Selected stake expired {self.unlock_datetime}.')
+        status = self.status()
+        if not status.is_child(Stake.Status.EDITABLE):
```
Comment (KPrasch) [inline nucypher/blockchain/eth/token.py:529]: Beautiful!

File: nucypher/cli/commands/stake.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/cli/commands/stake.py:702]: slow poke!

File: tests/integration/cli/actions/test_select_client_account.py [hunk 8]
```diff
@@ -156,8 +157,8 @@ def test_select_client_account_valid_sources(mocker,
         (1, True, True, True, []),
         (5, True, True, True, []),
         (NUMBER_OF_ETH_TEST_ACCOUNTS-1, True, True, True, []),
-        (4, True, True, True, [(1, 2, 3)]),
-        (7, True, True, True, [(1, 2, 3), (1, 2, 3)]),
+        (4, True, True, True, [SubStakeInfo(1, 2, 3)]),
```
Comment (KPrasch) [inline tests/integration/cli/actions/test_select_client_account.py:160]: Well... that was quite the dig to get here.

File: tests/integration/cli/actions/test_select_stake.py [hunk 9]
```diff
@@ -83,50 +126,75 @@ def assert_stake_table_not_painted(output: str) -> None:
...
-                                            mock_testerchain,
-                                            mock_stdin, # used to assert user hasn't been prompted
-                                            capsys):
+@pytest.mark.parametrize('sub_stakes_functions', [
+    [empty_sub_stakes],
+    [inactive_sub_stakes],
+    [unlocked_sub_stakes],
+    [not_editable_sub_stakes],
+    [inactive_sub_stakes, unlocked_sub_stakes, not_editable_sub_stakes]
+])
```
Comment (KPrasch) [inline tests/integration/cli/actions/test_select_stake.py:135]: cool!

File: nucypher/blockchain/eth/token.py (hunk 6 above)
Comment (vzotova) [reply nucypher/blockchain/eth/token.py:425]: `_status` is only for caching purposes, so it's more not yet calculated status

File: nucypher/blockchain/eth/token.py (hunk 5 above)
Comment (vzotova) [reply nucypher/blockchain/eth/token.py:282]: ✔️

File: nucypher/blockchain/eth/token.py [hunk 10]
```diff
@@ -249,43 +269,65 @@ def address_index_ordering_key(self):
...
     #
 
-    @property
-    def is_expired(self) -> bool:
-        current_period = self.staking_agent.get_current_period()  # TODO #1514 this is online only.
-        return bool(current_period > self.final_locked_period)
+    def status(self, staker_info: StakerInfo = None, current_period: Period = None) -> Status:
+        """
+        Returns status of sub-stake:
+        UNLOCKED - final period in a past
```
Comment (derekpierre) [inline nucypher/blockchain/eth/token.py:275]:
```suggestion
        UNLOCKED - final period in the past
```

File: nucypher/blockchain/eth/token.py [hunk 11]
```diff
@@ -249,43 +269,65 @@ def address_index_ordering_key(self):
...
+        INACTIVE - UNLOCKED and sub-stake will not be included in any future calculations
+        LOCKED - sub-stake is still locked and final period is current period
+        EDITABLE - LOCKED and final period greater than current
+        DIVISIBLE - EDITABLE and locked value is greater than two times of minimum allowed locked
```
Comment (derekpierre) [inline nucypher/blockchain/eth/token.py:279]:
```suggestion
        DIVISIBLE - EDITABLE and locked value is greater than two times the minimum allowed locked
```

File: nucypher/blockchain/eth/token.py (hunk 10 above)
Comment (vzotova) [reply nucypher/blockchain/eth/token.py]: ✔️

File: nucypher/blockchain/eth/token.py (hunk 11 above)
Comment (vzotova) [reply nucypher/blockchain/eth/token.py]: ✔️

File: nucypher/blockchain/eth/actors.py [hunk 12]
```diff
@@ -1674,21 +1665,20 @@ def assimilate(self, checksum_address: str, password: str = None) -> None:
 
         self.log.info(f"Resistance is futile - Assimilating Staker {original_form} -> {new_form}.")
```
Comment (cygnusv) [inline nucypher/blockchain/eth/actors.py:1666]: Perhaps it's a good moment to change this log message to something more in line with the current method. 
Something like this?
```suggestion
        self.log.info(f"Setting Staker from {original_form} to {new_form}.")
```

File: nucypher/blockchain/eth/token.py [hunk 13]
```diff
@@ -404,8 +445,10 @@ def divide(self, target_value: NU, additional_periods: int = None) -> Tuple['Sta
         self.sync()
 
         # Ensure selected stake is active
-        if self.is_expired:
-            raise self.StakingError(f'Cannot divide an expired stake. Selected stake expired {self.unlock_datetime}.')
+        status = self.status()
+        if not status.is_child(Stake.Status.DIVISIBLE):
```
Comment (cygnusv) [inline nucypher/blockchain/eth/token.py:449]: Maybe you can create a `is_divisible` property in `Status`. Same for `is_editable`.

File: nucypher/cli/painting/staking.py [hunk 14]
```diff
@@ -103,7 +107,6 @@ def prettify_stake(stake, index: int = None) -> str:
 
     pretty = f'| {index if index is not None else "-"} ' \
              f'| {stake.staker_address[:6]} ' \
-             f'| {stake.worker_address[:6]} ' \
```
Comment (cygnusv) [inline nucypher/cli/painting/staking.py:106]: Totally agree, it doesn't have sense to show the worker here, since all stakes have the same.

File: nucypher/blockchain/eth/token.py (hunk 13 above)
Comment (vzotova) [reply nucypher/blockchain/eth/token.py:449]: It was there but `Status` is for removing these methods, now we can describe sub-stake just with one status variable, otherwise will be needed to create flags for searching like it was before, and for each search - we need aditional flag. So I see that `Status` reduces code

File: nucypher/blockchain/eth/actors.py (hunk 12 above)
Comment (vzotova) [reply nucypher/blockchain/eth/actors.py]: ✔️

File: nucypher/blockchain/eth/token.py (hunk 4 above)
Comment (tuxxy) [reply nucypher/blockchain/eth/token.py:174]: `Enum`s are nice :rocket:

File: tests/integration/blockchain/test_token_and_stake.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    assert not Stake.Status.EDITABLE.is_child(Stake.Status.INACTIVE)
+    assert not Stake.Status.EDITABLE.is_child(Stake.Status.UNLOCKED)
+    assert Stake.Status.EDITABLE.is_child(Stake.Status.LOCKED)
+    assert not Stake.Status.EDITABLE.is_child(Stake.Status.DIVISIBLE)
+
+    # Check relations for divisible status
+    assert not Stake.Status.DIVISIBLE.is_child(Stake.Status.INACTIVE)
+    assert not Stake.Status.DIVISIBLE.is_child(Stake.Status.UNLOCKED)
+    assert Stake.Status.DIVISIBLE.is_child(Stake.Status.LOCKED)
+    assert Stake.Status.DIVISIBLE.is_child(Stake.Status.EDITABLE)
```
Comment (tuxxy) [inline tests/integration/blockchain/test_token_and_stake.py:56]: :+1:

File: nucypher/blockchain/eth/token.py [hunk 16]
```diff
@@ -159,6 +161,24 @@ class Stake:
...
+        EDITABLE = 4   # Editable
+        DIVISIBLE = 5  # Editable and divisible
+
+        def is_child(self, other: 'Status') -> bool:
+            if other == self.INACTIVE:
+                return self == self.INACTIVE
+            elif other == self.UNLOCKED:
+                return self.value <= self.UNLOCKED.value
+            else:
+                return self.value >= other.value
```
Comment (tuxxy) [inline nucypher/blockchain/eth/token.py:180]: I find this logic a bit hard to follow and weird for an `Enum`. You can simplify it by using an `IntEnum` rather than an `Enum` which confers `int`-like operations on the class itself.

Then it can be simplified to (I think):
```python
return self >= other or self < (self.UNLOCKED and self)
```

Or if you don't like condensed logic:
```python
if not other == self.UNLOCKED:
    return self >= other
return self < self.UNLOCKED
```

Comment (vzotova) [reply nucypher/blockchain/eth/token.py:180]: I'm not sure, in that case `self.INACTIVE == 1` will be `True` which looks weird for that `Status`

Comment (tuxxy) [reply nucypher/blockchain/eth/token.py:180]: In that case, you do `a_status.name`.

Comment (vzotova) [reply nucypher/blockchain/eth/token.py:180]: I understand, I mean `Status` instance will behave not as some class but as `int` which is not great for me
